test/jdk/jdk/crac/recompiler/NaturalDecompilationTest.java

```diff
@@ -0,0 +1,171 @@
+    private static final long STAGE_TIME_LIMIT_SEC = 30;
+
+    // The compilers have different deoptimization implementations so it is
+    // worth to test each of them.
+    public enum Variant {
+        TIERED,
+        C1_ONLY,
+        C2_ONLY,
+    };
+
+    @CracTestArg
+    Variant variant;
+
+    private static void blackhole(@SuppressWarnings("unused") Object o) {}
+
+    private static int testMethod(int i) {
+        try {
+            // A compiler intrinsic which triggers decompilation in both C1 and
+            // C2 when it has to throw IndexOutOfBoundsException
+            Objects.checkIndex(i, 10);
+            return i;
+        } catch (IndexOutOfBoundsException ignored) {
+            return -1;
+        }
+    }
+
+    private static final String BLACKHOLE_NAME = "blackhole";
+    private static final String TEST_METHOD_NAME = "testMethod";
+
+    private static final int TEST_ARG_EXPECTED = 0;
+    private static final int TEST_ARG_UNEXPECTED = -1;
+
+    @Override
+    public void test() throws Exception {
+        final var builder = new CracBuilder().engine(CracEngine.SIMULATE).captureOutput(true)
+            .vmOption("-Xbootclasspath/a:wb.jar").vmOption("-XX:+UnlockDiagnosticVMOptions").vmOption("-XX:+WhiteBoxAPI")
+            .javaOption("jdk.crac.recompilation-delay-ms", "0")
+            .vmOption("-XX:+UnlockExperimentalVMOptions")
+            .vmOption("-XX:CompileCommand=blackhole," + NaturalDecompilationTest.class.getName() + "." + BLACKHOLE_NAME)
+            .vmOption("-XX:CompileCommand=dontinline," + NaturalDecompilationTest.class.getName() + "." + TEST_METHOD_NAME)
+            .vmOption("-XX:+PrintCompilation")
+            .vmOption("-Xlog:crac=trace");
+        switch (variant) {
+            case TIERED -> { /* This is the default */ }
+            case C1_ONLY -> builder.vmOption("-XX:TieredStopAtLevel=1");
+            case C2_ONLY -> builder.vmOption("-XX:-TieredCompilation");
+        }
+
+        // Must create an output analyzer before waiting for success, otherwise
+        // on Windows PrintCompilation overflows the piping buffer and the
+        // waiting never completes
+        final var proc = builder.startCheckpoint();
+        final var out = proc.outputAnalyzer();
+        proc.waitForSuccess();
+        out.shouldContain("Requesting recompilation: int " + NaturalDecompilationTest.class.getName() + "." + TEST_METHOD_NAME + "(int)");
+    }
+
+    @Override
+    public void exec() throws Exception {
+        final var whiteBox = WhiteBox.getWhiteBox();
+        final var testMethodRef = NaturalDecompilationTest.class.getDeclaredMethod(TEST_METHOD_NAME, int.class);
+
+        timedDoWhile("compilation", () -> {
+            for (int i = 0; i < 2000; i++) {
+                final var res = testMethod(TEST_ARG_EXPECTED);
+                blackhole(res);
+            }
+            try {
+                Thread.sleep(500); // Time to compile
```

Comment:
I think this has some potential for false positives in the testsuite. I guess there is no API in whitebox to wait until this is observed? Alternatively watching the output in `test()` and signaling the process somehow...